feat(permissions): system user support for permission check v2

kkrime opened the pull request:
# Which Problems Are Solved

For permissoin check v2, we currently check the human users permissions from the DB, however for system users (from the config, which would not be in the DB), their permissions are not in the DB, but are in a config file like defaults.yaml.

- The current check involves **all** users (human, machine) in the DB and excludes only system users
- It's redundant to mention twice that system users are not in the DB.

 ## How the Problems Are Solved

- Can you please expand a bit on your implementation? For example:
  - what is the precedence of the permissions being applied.
  - which role-permission configuration is used for the system users (should be runtime, but currently is DB, see my other comment)
  - Mention details such as "roles are loaded once per request and cached", which you did in a separate comment.

The way the permissions work are based on the following: 
https://zitadel.slack.com/archives/C087ADF8LRX/p1742207808062949?thread_ts=1742206770.965909&cid=C087ADF8LRX

# Additional Changes

**Important to note** from this point onwards, System Users permissions will be picked up from `SystemAuthz` (added to `default.yaml`) **NOT** `InternalAuthz`


- Closes https://github.com/zitadel/zitadel/issues/9189

Commit:
feat(permissions): Addeding system user support for permission check v2

Commit:
fixup! feat(permissions): Addeding system user support for permission check v2

Commit:
fixup! fixup! feat(permissions): Addeding system user support for permission check v2

Commit:
fixup! fixup! fixup! feat(permissions): Addeding system user support for permission check v2

Commit:
fixup! fixup! fixup! fixup! feat(permissions): Addeding system user support for permission check v2

Commit:
fixup! fixup! fixup! fixup! fixup! feat(permissions): Addeding system user support for permission check v2

Commit:
fixup! fixup! fixup! fixup! fixup! fixup! feat(permissions): Addeding system user support for permission check v2

Commit:
fixup! fixup! fixup! fixup! fixup! fixup! fixup! feat(permissions): Addeding system user support for permission check v2

Commit:
fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! feat(permissions): Addeding system user support for permission check v2

Commit:
fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! feat(permissions): Addeding system user support for permission check v2

Commit:
fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! feat(permissions): Addeding system user support for permission check v2

Commit:
fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! feat(permissions): Addeding system user support for permission check v2

Commit:
fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! feat(permissions): Addeding system user support for permission check v2

kkrime commented on internal/api/authz/authorization.go:
```diff
@@ -125,3 +128,47 @@ func GetAllPermissionCtxIDs(perms []string) []string {
+func addGetSystemRolesFuncToCtx(ctx context.Context, ctxData CtxData) context.Context {
+	if len(ctxData.SystemMemberships) != 0 {
+		ctx = context.WithValue(ctx, systemUserRolesFuncKey, func() func(ctx context.Context) ([]string, error) {
+			var roles []string
```
My thinking was, there might be more complex API calls that require multiple DB calls, so I implemented it in this way using `var roles []string` as a cache

If my assumption is wrong, let me know and I'll get rid of this and simplfy things

muhlemmer replied:
Another note, a `context.Context` is supposed to be routine safe. The `roles` variable will have a race condition if a system user context is ever passed to multiple Go routines at once.

I think it will be safer and simpler to use a [`sync.OnceFunc`](https://pkg.go.dev/sync#OnceFunc) instead.

Commit:
fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! feat(permissions): Addeding system user support for permission check v2

Commit:
fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! feat(permissions): Addeding system user support for permission check v2

Commit:
fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! feat(permissions): Addeding system user support for permission check v2

Commit:
fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! feat(permissions): Addeding system user support for permission check v2

muhlemmer commented on internal/api/authz/authorization.go:
```diff
@@ -125,3 +128,47 @@ func GetAllPermissionCtxIDs(perms []string) []string {
+	}
+	getSystemUserRolesFunc, ok := getSystemUserRolesFuncValue.(func(context.Context) ([]string, error))
+	if !ok {
+		return nil, errors.New("unable to obtain systems role func")
```
I'm wondering if it's really needed to return an error here. What will fail if there are no system membership entries or function? Can't we just return a empty value and only log the error of type-mismatch? Technically type mis-match would be a bug and shouldn't really get exposed on the API.

muhlemmer commented on internal/api/authz/authorization.go:
```diff
@@ -125,3 +128,47 @@ func GetAllPermissionCtxIDs(perms []string) []string {
+func getSystemUserRoles(ctx context.Context) ([]string, error) {
+	ctxData, ok := ctx.Value(dataKey).(CtxData)
+	if !ok {
+		return nil, errors.New("unable to obtain ctxData")
```
See other error comment

muhlemmer commented on cmd/setup/51/01-permitted_orgs_function.sql:
```diff
@@ -0,0 +1,77 @@
+      permission_found_in_system_roles bool;
+    BEGIN
+      SELECT result.role_found INTO permission_found_in_system_roles
+      FROM (SELECT matched_roles && system_user_roles AS role_found) AS result;
```
For the system user we should only use the static role-permission definition from the runtime configuration. When using the instance level role-permissions, it is will be [possible in the future](https://github.com/zitadel/zitadel/issues/9189) to change the definition of the roles and effectively be able to lockout system users.

From the linked issue's acceptance criteria:

> Optional arguments are passed to the permission functions with the system user memberships and role permission mappings.

Commit:
fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! feat(permissions): Addeding system user support for permission check v2

Commit:
fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! feat(permissions): Addeding system user support for permission check v2

Commit:
fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! feat(permissions): Addeding system user support for permission check v2

Commit:
fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! feat(permissions): Addeding system user support for permission check v2

Commit:
fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! feat(permissions): Addeding system user support for permission check v2

Commit:
fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! feat(permissions): Addeding system user support for permission check v2

Commit:
fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! feat(permissions): Addeding system user support for permission check v2

Commit:
fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! feat(permissions): Addeding system user support for permission check v2

Commit:
fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! feat(permissions): Addeding system user support for permission check v2

Commit:
fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! feat(permissions): Addeding system user support for permission check v2

Commit:
fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! feat(permissions): Addeding system user support for permission check v2

Commit:
fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! feat(permissions): Addeding system user support for permission check v2

Commit:
fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! feat(permissions): Addeding system user support for permission check v2

Commit:
fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! feat(permissions): Addeding system user support for permission check v2

Commit:
fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! feat(permissions): Addeding system user support for permission check v2

Commit:
fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! feat(permissions): Addeding system user support for permission check v2

Commit:
fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! feat(permissions): Addeding system user support for permission check v2

Commit:
fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! feat(permissions): Addeding system user support for permission check v2

Commit:
Merge branch 'main' into syste-users-permissions

Commit:
fixup! Merge branch 'main' into syste-users-permissions

Commit:
fixup! fixup! Merge branch 'main' into syste-users-permissions

Commit:
fixup! fixup! fixup! Merge branch 'main' into syste-users-permissions

Commit:
fixup! fixup! fixup! fixup! Merge branch 'main' into syste-users-permissions

Commit:
fixup! fixup! fixup! fixup! fixup! Merge branch 'main' into syste-users-permissions

Commit:
fixup! fixup! fixup! fixup! fixup! fixup! Merge branch 'main' into syste-users-permissions

muhlemmer commented on cmd/mirror/config.go:
```diff
@@ -28,10 +28,8 @@ type Migration struct {
-var (
-	//go:embed defaults.yaml
-	defaultConfig []byte
-)
+//go:embed defaults.yaml
+var defaultConfig []byte
 
```
unrelated change? please revert

muhlemmer commented on cmd/mirror/projections.go:
```diff
@@ -184,7 +184,7 @@ func projections(
 		&http.Client{},
 		func(ctx context.Context, permission, orgID, resourceID string) (err error) {
-			return internal_authz.CheckPermission(ctx, authZRepo, config.InternalAuthZ.RolePermissionMappings, permission, orgID, resourceID)
+			return internal_authz.CheckPermission(ctx, authZRepo, nil, config.InternalAuthZ.RolePermissionMappings, permission, orgID, resourceID)
```
Is the systemUserRoleMapping argument supposed to stay `nil`?

muhlemmer reviewed (commented):
Nice work. I left a number of small comments.

I wasn't able to do manual testing, yet. I'll do that once the comments are resolved.

kkrime commented on internal/api/authz/authorization.go:
```diff
@@ -125,3 +129,63 @@ func GetAllPermissionCtxIDs(perms []string) []string {
+					return systemUserPermissionsDbJsonQueryStruct
+				}
+
+				chann <- struct{}{}
+				defer func() {
+					<-chann
+				}()
```
Ya, this relates to your previous comment; https://github.com/zitadel/zitadel/pull/9460#discussion_r1989314374

I couldn't use `[sync.OnceFunc](https://pkg.go.dev/sync#OnceFunc)` becuase it only supports `func()` with no parameters, so I used this instead

kkrime commented on cmd/setup/52/02-permitted_orgs_function.sql:
```diff
@@ -0,0 +1,162 @@
+      check_aggregates bool;
+    BEGIN
+
+    DROP TABLE IF EXISTS matching_system_user_perms;
+    CREATE TEMPORARY TABLE matching_system_user_perms (
+            member_type TEXT,
+            aggregate_id TEXT,
+            object_id TEXT
+        ) ON COMMIT DROP;
```
I looked into this, CTEs are part of a single statement/query, I woudln't be able to use the CTE throughout the funciton. 
If I'm wrong on this, let me know

muhlemmer replied:
As the function stands now, indeed a CTE won't work. However, I am still not comfortable using a temporary table. It will fill the system catalogs on every request, which incurs some disk writes as well.

Currently the function is mostly a conditional ladder. Perhaps we compact some parts of the function by using a custom on the membership type, or a simple `CASE` switch. 

Custom order can be achieved with something like:

```sql
ORDER BY
	array_position(array[
		'System'
		, 'IAM'
		, 'Organization'
		, 'Project'
		, 'ProjectGrant'
	], m.member_type)
```

kkrime committed:
fixup! fixup! fixup! fixup! fixup! fixup! fixup! Merge branch 'main' into syste-users-permissions

Tims requested changes

kkrime committed:
fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! Merge branch 'main' into syste-users-permissions

muhlemmer commented on internal/query/user.go:
```diff
@@ -655,7 +655,10 @@ func (q *Queries) searchUsers(ctx context.Context, queries *UserSearchQueries, f
-		query = wherePermittedOrgsOrCurrentUser(ctx, query, filterOrgIds, UserResourceOwnerCol.identifier(), UserIDCol.identifier(), domain.PermissionUserRead)
+		query, err = wherePermittedOrgsOrCurrentUser(ctx, query, filterOrgIds, UserResourceOwnerCol.identifier(), UserIDCol.identifier(), domain.PermissionUserRead)
+		if err != nil {
+			return nil, zerrors.ThrowInternal(err, "AUTHZ-HS4us", "Errors.Internal")
```
1. I believe the internal error is now thrown twice. Once in `wherePermittedOrgsOrCurrentUser` and again here with the parent wrapped. Please keep only one of them. Make you choice will be consistent in `wherePermittedOrgs`
2. For the error ID, use `QUERY-<random string>`. `AUTHZ` is typically used by the `authz` package I would say.

muhlemmer reviewed (commented):
Starting to look really good :). Just 1 more small comment.

I left the CTE and sync.Once comments open for a moment, I will quickly check feasability on my side to make sure I'm not talking BS. If there's something to improve there I will send a PR against your branch. Otherwise I'll be okay with closing the comments & merge.

muhlemmer commented:
Also note there are still failing pipelines

muhlemmer committed:
fix(authz): simplify system permission mapping (#9627)

# Which Problems Are Solved

Remove the need of maintaining a channel for System user permission
mapping, while still ensuring the mapping is built only once per
request.

# How the Problems Are Solved

Build the slice of `SystemUserPermissionsDBQuery` once, when the context
is being created in the context middleware. This slice will remain valid
during the request lifetime as part of the context. This removes the
need of "caching" and additional synchronization.

# Additional Changes

```sh
 gci write . --skip-generated -s standard -s default -s 'prefix(github.com/zitadel/zitadel)' --custom-order
```

# Additional Context

- Follow up on discussion
https://github.com/zitadel/zitadel/pull/9460#discussion_r2006769497

kkrime committed:
Merge branch 'main' into syste-users-permissions

kkrime committed:
Merge branch 'main' into syste-users-permissions

kkrime committed:
fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! Merge branch 'main' into syste-users-permissions

kkrime commented on internal/query/permission.go:
```diff
@@ -2,51 +2,74 @@ package query
-	userID := authz.GetCtxData(ctx).UserID
-	logging.WithFields("permission_check_v2_flag", authz.GetFeatures(ctx).PermissionCheckV2, "org_id_column", orgIDColumn, "permission", permission, "user_id", userID).Debug("permitted orgs check used")
+// func wherePermittedOrgs(ctx context.Context, query sq.SelectBuilder, filterOrgIds, orgIDColumn, permission string) (sq.SelectBuilder, error) {
+// 	userID := authz.GetCtxData(ctx).UserID
```
@muhlemmer I had to comment this out to pass the Linting pipeline

kkrime commented on e2e/config/localhost/docker-compose.yaml:
```diff
@@ -30,7 +30,7 @@ services:
     restart: 'always'
-    image: 'cockroachdb/cockroach:latest-v24.3'
+    image: 'cockroachdb/cockroach:latest'
     command: 'start-single-node --insecure --http-addr :9090'
```
@livio-a I need to make this change otherwise it starts complaining about a unsupported feture in my sql scripts

livio-a replied:
if it doesn't work with cockroach, we need to make sure it's only released as part of v3 (target branch for that is `v3.x`)

kkrime committed:
fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! Merge branch 'main' into syste-users-permissions

kkrime committed:
replaced cockroach with postgres

kkrime committed:
Revert "replaced cockroach with postgres"

This reverts commit 96cf271fb8c271cf49cfef016308451f6c08bc53.

kkrime committed:
trying cte

kkrime committed:
fixup! trying cte

kkrime committed:
fixup! fixup! trying cte

kkrime committed:
fixup! fixup! fixup! trying cte

kkrime committed:
Reapply "replaced cockroach with postgres"

This reverts commit 4bc7b0a6877987d0351e3ff35677d1ebeb6b50fa.

kkrime committed:
fixup! Reapply "replaced cockroach with postgres"

kkrime committed:
Merge branch 'main' into syste-users-permissions

kkrime committed:
removed unused variables in eventstore.check_system_user_perms

muhlemmer commented:
Replaced by https://github.com/zitadel/zitadel/pull/9640
